### ISSUE TITLE
`SearchLight.save!! ` not working with GenieAuthentication 

### ISSUE THREAD

**Issue by cihga39871** [GenieFramework/SearchLightSQLite.jl#3]
Hi,

I created a full-stack project using SQLite database, and adds GenieAuthentication plugin, but I found an error when register for a new user.

In the file AuthenticationController.jl:
```julia
user = SearchLight.save!!(User( username  = Genie.Router.@params(:username),
                                    password  = Genie.Router.@params(:password) |> Users.hash_password,
                                    name      = Genie.Router.@params(:name),
                                    email     = Genie.Router.@params(:email)))
```

It will throw the error:
```
ERROR: MethodError: Cannot `convert` an object of type Array{SQLInput,1} to an object of type SQLInput
Closest candidates are:
  convert(::Type{SQLInput}, ::Real) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:104
  convert(::Type{SQLInput}, ::Symbol) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:105
  convert(::Type{SQLInput}, ::Dates.DateTime) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:106
  ...
Stacktrace:
 [1] to_sqlinput(::Users.User, ::Symbol, ::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:607
 [2] (::SearchLightSQLite.var"#4#6"{Users.User})(::String) at /home/jc/.julia/packages/SearchLightSQLite/UKD3j/src/SearchLightSQLite.jl:207
 [3] iterate at ./generator.jl:47 [inlined]
 [4] collect_to!(::Array{String,1}, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Int64, ::Int64) at ./array.jl:711
 [5] collect_to_with_first!(::Array{String,1}, ::String, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Int64) at ./array.jl:689
 [6] _collect(::Array{String,1}, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./array.jl:683
 [7] collect_similar at ./array.jl:607 [inlined]
 [8] map at ./abstractarray.jl:2072 [inlined]
 [9] to_store_sql(::Users.User; conflict_strategy::Symbol) at /home/jc/.julia/packages/SearchLightSQLite/UKD3j/src/SearchLightSQLite.jl:207
 [10] _save!!(::Users.User; conflict_strategy::Symbol, skip_validation::Bool, skip_callbacks::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:190
 [11] save!!(::Users.User; conflict_strategy::Symbol, skip_validation::Bool, skip_callbacks::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:160
 [12] save!!(::Users.User) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:160
```

# Replication:
## Creating PolyChromeGenie Project

In Julia v1.4 or v1.5.1 repl:

```julia
using Genie, SearchLight
Genie.newapp_fullstack("PolyChromeGenie")
```

## Setting up Database

https://genieframework.github.io/Genie.jl/dev/guides/Working_With_Genie_Apps.html#Accessing-databases-with-SearchLight-models

After starting `bin/repl`:

```julia
julia> ]
(PolyChromeGenie) pkg> add SearchLightSQLite
```

Edit `db/connection.yml`:

```
env: ENV["GENIE_ENV"]

dev:
  adapter: SQLite
  database: db/polychromegenie.sqlite
```

Connect the database `db/polychromegenie.sqlite` in Genie repl:

```julia
using SearchLight, SearchLightSQLite
SearchLight.Configuration.load() |> SearchLight.connect

# create migration table for later use, such as Authentication plugins
SearchLight.Migrations.create_migrations_table()
```

## Add authentication plugin

https://github.com/GenieFramework/GenieAuthentication.jl

In Genie repl:

```julia
(PolyChromeGenie) pkg> add GenieAuthentication

julia> using GenieAuthentication
julia> GenieAuthentication.install(@__DIR__)

# Setup database
julia> using SearchLight
julia> SearchLight.Migration.up("CreateTableUsers")
```

### Add ViewHelper module to display auth result

In `app/resources/authentication/AuthenticationController.jl`, add the line:

```julia
using ViewHelper
```

which fix the error: cannot find `output_flash()` in the views of `login.jl.html` and `register.jl.html`

### Enable registration

Uncomment the register lines in `plugins/genie_authentication.jl`.

### Fix bug of throw errors without the `:msg` field

In the `register()`function in `app/resources/authentication/AuthenticationController.jl`,

change `Genie.Flash.flash(ex.msg)` to:

```julia
if hasfield(typeof(ex), :msg)
  Genie.Flash.flash(ex.msg)
else
  rethrow(ex)
end
```

## Bug here
In REPL, run `up()`, and go to the web page `http://127.0.0.1:8000/register`, and create a user, the error will display:
```
500 Internal Error - MethodError: Cannot `convert` an object of type Array{SearchLight.SQLInput,1} to an object of type SearchLight.SQLInput
Closest candidates are:
  convert(::Type{SearchLight.SQLInput}, !Matched::Real) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:104
  convert(::Type{SearchLight.SQLInput}, !Matched::Symbol) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:105
  convert(::Type{SearchLight.SQLInput}, !Matched::Dates.DateTime) at /home/jc/.julia/packages/SearchLight/lFwBS/src/model_types.jl:106
  ...
to_sqlinput(::Users.User, ::Symbol, ::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:607
(::SearchLightSQLite.var"#4#6"{Users.User})(::String) at /home/jc/.julia/packages/SearchLightSQLite/UKD3j/src/SearchLightSQLite.jl:207
iterate at ./generator.jl:47 [inlined]
collect_to!(::Array{String,1}, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Int64, ::Int64) at ./array.jl:732
collect_to_with_first!(::Array{String,1}, ::String, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Int64) at ./array.jl:710
_collect(::Array{String,1}, ::Base.Generator{Array{String,1},SearchLightSQLite.var"#4#6"{Users.User}}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./array.jl:704
collect_similar at ./array.jl:628 [inlined]
map at ./abstractarray.jl:2162 [inlined]
to_store_sql(::Users.User; conflict_strategy::Symbol) at /home/jc/.julia/packages/SearchLightSQLite/UKD3j/src/SearchLightSQLite.jl:207
_save!!(::Users.User; conflict_strategy::Symbol, skip_validation::Bool, skip_callbacks::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:190
save!!(::Users.User; conflict_strategy::Symbol, skip_validation::Bool, skip_callbacks::Array{Symbol,1}) at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:160
save!! at /home/jc/.julia/packages/SearchLight/lFwBS/src/SearchLight.jl:160 [inlined]
register() at /home/jc/projects/PolyChromeGUI/PolyChromeGenie/app/resources/authentication/AuthenticationController.jl:41
#invokelatest#1 at ./essentials.jl:710 [inlined]
invokelatest at ./essentials.jl:709 [inlined]
match_routes(::HTTP.Messages.Request, ::HTTP.Messages.Response, ::Genie.Router.Params{Any}) at /home/jc/.julia/packages/Genie/iL6x9/src/Router.jl:502
route_request(::HTTP.Messages.Request, ::HTTP.Messages.Response, ::Sockets.IPv4) at /home/jc/.julia/packages/Genie/iL6x9/src/Router.jl:150
handle_request(::HTTP.Messages.Request, ::HTTP.Messages.Response, ::Sockets.IPv4) at /home/jc/.julia/packages/Genie/iL6x9/src/AppServer.jl:168
handle_request at /home/jc/.julia/packages/Genie/iL6x9/src/AppServer.jl:161 [inlined]
#22 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/macros.jl:127 [inlined]
run_work_thunk(::Genie.AppServer.var"#22#23"{HTTP.Messages.Request,HTTP.Messages.Response}, ::Bool) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/process_messages.jl:79
remotecall_fetch(::Function, ::Distributed.LocalProcess; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/remotecall.jl:379
remotecall_fetch(::Function, ::Distributed.LocalProcess) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/remotecall.jl:379
remotecall_fetch(::Function, ::Int64; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/remotecall.jl:421
remotecall_fetch(::Function, ::Int64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Distributed/src/remotecall.jl:421
setup_http_handler(::HTTP.Messages.Request, ::HTTP.Messages.Response) at /home/jc/.julia/packages/Genie/iL6x9/src/AppServer.jl:183
setup_http_handler at /home/jc/.julia/packages/Genie/iL6x9/src/AppServer.jl:182 [inlined]
handle at /home/jc/.julia/packages/HTTP/IAI92/src/Handlers.jl:253 [inlined]
handle(::HTTP.Handlers.RequestHandlerFunction{typeof(Genie.AppServer.setup_http_handler)}, ::HTTP.Streams.Stream{HTTP.Messages.Request,HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}}) at /home/jc/.julia/packages/HTTP/IAI92/src/Handlers.jl:276
(::Genie.AppServer.var"#6#13"{Int64,Int64})(::HTTP.Streams.Stream{HTTP.Messages.Request,HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}}) at /home/jc/.julia/packages/Genie/iL6x9/src/AppServer.jl:78
macro expansion at /home/jc/.julia/packages/HTTP/IAI92/src/Servers.jl:367 [inlined]
(::HTTP.Servers.var"#13#14"{Genie.AppServer.var"#6#13"{Int64,Int64},HTTP.ConnectionPool.Transaction{Sockets.TCPSocket},HTTP.Streams.Stream{HTTP.Messages.Request,HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}}})() at ./task.jl:356
```

## My Environment:
```julia
(PolyChromeGenie) pkg> status
Status `~/projects/PolyChromeGUI/PolyChromeGenie/Project.toml`
  [c43c736e] Genie v1.7.1
  [e115e502] GenieAuthentication v0.4.2
  [e6f89c97] LoggingExtras v0.4.2
  [739be429] MbedTLS v1.0.2
  [340e8cb6] SearchLight v0.20.1
  [21a827c4] SearchLightSQLite v0.4.1
```

**Comment by essenciary** [comment]
@cihga39871 Thanks for reporting this! I'll check it out ASAP.

**Comment by essenciary** [comment]
@cihga39871 Thanks for all the details to reproduce the problem. I can't figure out what happened but something was badly borked up in git/releases. 

The problem is the `Users.jl` model which has been updated months ago and is no longer compatible with SearchLight. Somehow the changes were lost. 

You will need to replace the contents of the `Users.jl` file with what's here: 
https://github.com/GenieFramework/GenieAuthentication.jl/blob/master/files/app/resources/users/Users.jl

Also update to latest v0.5.

PS - I've incorporated your fixes as well into the controller, thanks!
